### PR TITLE
[libarchive,lz4] Fix pc file, modernize portfile

### LINKED ISSUE #20182: [libarchive] wrapper emits warnings and uses find_dependency
related issue about the warning:
https://gitlab.kitware.com/vtk/vtk/-/issues/18308

also the wrapper uses `find_dependency` instead of `find_package` which only works correctly in `*config.cmake`

### PR DESCRIPTION
**Pull request by dg0yt**
- #### What does your PR fix?  
  - [x] Fixes the invalid pkgconfig file by switching to modules. (Cf. https://github.com/microsoft/vcpkg/pull/20062#issuecomment-916377810.)
  - [x] Modernizes the portfile. Cleanup.
  - [x] Drops the unused openssl dependency for osx. 
  - [x] Fix `find_dependency` in the wrapper (#20182).

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  unchanged, no

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  yes

- #### Features
  pc file after `vcpkg install 'libarchive[*]'` on x64-osx:

~~~
prefix=${pcfiledir}/../..

exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/../include

Name: libarchive
Description: library that can create and read several streaming archive formats
Version: 3.5.2
Cflags: -I"${includedir}"
Cflags.private: -DLIBARCHIVE_STATIC
Libs: -L"${libdir}" -larchive 
Requires: zlib bzip2 liblzma lzo2 liblz4 libzstd libxml-2.0
~~~
 (Note that `libssl` and `libcrypto` are missing. That's how I found that libarchive doesn't use OpenSSL on 'Darwin'.)

### CODE REVIEW

File: ports/libarchive/portfile.cmake [hunk 1]
```diff
@@ -37,21 +38,14 @@ vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
         #libb2   ENABLE_LIBB2
```
Comment (NancyLi1013) [inline ports/libarchive/portfile.cmake:38]: Since these features are commented out. We can remove these comments until we add these as features.

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by dg0yt** [comment]
Hm, there are no pc files for the MSVC builds of port openssl.

**Comment by dg0yt** [comment]
This is complete. If possible this should be tested with #20062 on windows, static and dynamic.

File: ports/libarchive/vcpkg-cmake-wrapper.cmake.in [hunk 2]
```diff
@@ -1,123 +1,63 @@
...
-            list(APPEND LibArchive_LIBRARIES debug ${LZO_LIBRARY_DEBUG})
+        include(SelectLibraryConfigurations)
+        find_library(Z_VCPKG_LZO_LIBRARY_DEBUG NAMES lzo2d lzo2 PATHS "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/debug/lib" NO_DEFAULT_PATH)
+        find_library(Z_VCPKG_LZO_LIBRARY_RELEASE NAMES lzo2 PATHS "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/lib" NO_DEFAULT_PATH)
+        select_library_configurations(Z_VCPKG_LZO)
+        if(Z_VCPKG_LZO_LIBRARY)
+            list(APPEND z_vcpkg_libarchive_libs ${Z_VCPKG_LZO_LIBRARY})
+        else()
+            set(LibArchive_FOUND FALSE)
         endif()
```
Comment (Neumann-A) [inline ports/libarchive/vcpkg-cmake-wrapper.cmake.in:41]: Why `Z_VCPKG` ? Where is the conflict?

Comment (dg0yt) [reply ports/libarchive/vcpkg-cmake-wrapper.cmake.in:41]: Well, it is a wrapper. It is run in user project scope, isn't it?. And `select_library_configurations` is know to set `<PREFIX>_FOUND`. So this just tries to go out of the way of user projects.

**Comment by NancyLi1013** [comment]
The problem caused by CI pipeline line depends on PR https://github.com/microsoft/vcpkg/pull/20236.

**Comment by c72578** [comment]
Remark: This is how **`libarchive.pc`** looks now (at 141ec0d4ccf3a65272c284ef49d573c8e696fc48) in case of `x64-windows`:

```
prefix=${pcfiledir}/../..

exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include

Name: libarchive
Description: library that can create and read several streaming archive formats
Version: 3.5.2
Cflags: -I"${includedir}"
Cflags.private: -DLIBARCHIVE_STATIC
Libs: -L"${libdir}" -larchive
Libs.private:  -lcrypto -lssl -lssl -lssl -lssl -lssl -lssl
Requires.private:  zlib bzip2 liblzma liblz4 libzstd libxml-2.0
```
It contains `-lssl` 6 times under `Libs.private`

**Comment by dg0yt** [comment]
Thanks for the feedback @c72578. I assume it works now for #20062?

> It contains `-lssl` 6 times under `Libs.private`

I don't think this is a problem for now. The library is added for each matching algorithm. This could be improved upstream, instead of adding complexity to the patch. Maybe when I upstream the patch...

**Comment by c72578** [comment]
@dg0yt the situation concerning #20062 has improved, however it is still failing for **`x64-linux`**:
`/usr/bin/ld: cannot find -llz4`

This is how **`libarchive.pc`** looks under **`x64-linux`** now (tested at commit 74f3712):
```
prefix=${pcfiledir}/../..

exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include

Name: libarchive
Description: library that can create and read several streaming archive formats
Version: 3.5.2
Cflags: -I"${includedir}"
Cflags.private: -DLIBARCHIVE_STATIC
Libs: -L"${libdir}" -larchive 
Requires: zlib bzip2 liblzma liblz4 libzstd libcrypto libxml-2.0 libssl libssl libssl libssl libssl libssl
```
See: `liblz4` in `Requires`

**Comment by c72578** [comment]
@dg0yt The remaining issue is related to the port **`lz4`**.
`lz4_x64-linux/debug/lib/pkgconfig\liblz4.pc` contains `-llz4`, but it should contain `-llz4d`
```
prefix=${pcfiledir}/../..
#   LZ4 - Fast LZ compression algorithm
#   Copyright (C) 2011-2014, Yann Collet.
#   BSD 2-Clause License (http://www.opensource.org/licenses/bsd-license.php)

libdir=${prefix}/lib
includedir=${prefix}/../include

Name: lz4
Description: extremely fast lossless compression algorithm library
URL: http://www.lz4.org/
Version: 1.9.3
Libs: -L"${prefix}/lib" -llz4
Cflags: -I"${prefix}/../include"
```

**Comment by dg0yt** [comment]
> `lz4_x64-linux/debug/lib/pkgconfig\liblz4.pc` contains `-llz4`, but it should contain `-llz4d`

Thanks for reporting @c72578. I added a fix to this PR now.

**Review by c72578 — APPROVED** [review]
@dg0yt Thanks. I have completed the testing and the PR LGTM now.

**Comment by c72578** [comment]
@NancyLi1013 PR #20236 has been merged in the meantime.

**Review by NancyLi1013 — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks @dg0yt for the fixes, @c72578 @JonLiu1993 for testing, and @Neumann-A for the review!

File: ports/libarchive/pkgconfig-modules.patch [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+ #	 This only affects Libs.private (looked up for static builds) though.
+diff --git a/build/pkgconfig/libarchive.pc.in b/build/pkgconfig/libarchive.pc.in
+index 4b631e6..80198f1 100644
+--- a/build/pkgconfig/libarchive.pc.in
++++ b/build/pkgconfig/libarchive.pc.in
+@@ -10,3 +10,4 @@ Cflags: -I${includedir}
+ Cflags.private: -DLIBARCHIVE_STATIC
+ Libs: -L${libdir} -larchive
+ Libs.private: @LIBS@
++Requires.private: @LIBARCHIVE_REQUIRES_PRIVATE@
```
[the remark this answers is not in the document]
Comment (kientzle) [reply ports/libarchive/pkgconfig-modules.patch:238]: I would appreciate if someone could help us verify libarchive/libarchive#2253 which claims to be a fix for libarchive/libarchive#1446

**Pull request merged**